src/sdio.rs

```diff
@@ -0,0 +1,825 @@
+pub trait PinD2 {}
+pub trait PinD3 {}
+
+pub trait Pins {
+    const BUSWIDTH: Buswidth;
+}
+
+impl<CLK, CMD, D0, D1, D2, D3> Pins for (CLK, CMD, D0, D1, D2, D3)
+where
+    CLK: PinClk,
+    CMD: PinCmd,
+    D0: PinD0,
+    D1: PinD1,
+    D2: PinD2,
+    D3: PinD3,
+{
+    const BUSWIDTH: Buswidth = Buswidth::Buswidth4;
+}
+
+impl<CLK, CMD, D0> Pins for (CLK, CMD, D0)
+where
+    CLK: PinClk,
+    CMD: PinCmd,
+    D0: PinD0,
+{
+    const BUSWIDTH: Buswidth = Buswidth::Buswidth1;
+}
+
+macro_rules! pins {
+    ($(CLK: [$($CLK:ty),*] CMD: [$($CMD:ty),*] D0: [$($D0:ty),*] D1: [$($D1:ty),*] D2: [$($D2:ty),*] D3: [$($D3:ty),*])+) => {
+        $(
+            $(
+                impl PinClk for $CLK {}
+            )*
+            $(
+                impl PinCmd for $CMD {}
+            )*
+            $(
+                impl PinD0 for $D0 {}
+            )*
+            $(
+                impl PinD1 for $D1 {}
+            )*
+            $(
+                impl PinD2 for $D2 {}
+            )*
+            $(
+                impl PinD3 for $D3 {}
+            )*
+        )+
+    }
+}
+
+#[cfg(any(
+    feature = "stm32f401",
+    feature = "stm32f405",
+    feature = "stm32f407",
+    feature = "stm32f411",
+    feature = "stm32f412",
+    feature = "stm32f413",
+    feature = "stm32f415",
+    feature = "stm32f417",
+    feature = "stm32f423",
+    feature = "stm32f427",
+    feature = "stm32f429",
+    feature = "stm32f437",
+    feature = "stm32f439",
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479"
+))]
+pins! {
+    CLK: [PC12<Alternate<AF12>>]
+    CMD: [PD2<Alternate<AF12>>]
+    D0: [PC8<Alternate<AF12>>]
+    D1: [PC9<Alternate<AF12>>]
+    D2: [PC10<Alternate<AF12>>]
+    D3: [PC11<Alternate<AF12>>]
+}
+
+#[cfg(any(
+    feature = "stm32f411",
+    feature = "stm32f412",
+    feature = "stm32f413",
+    feature = "stm32f423",
+))]
+pins! {
+    CLK: [PB15<Alternate<AF12>>]
+    CMD: [PA6<Alternate<AF12>>]
+    D0: [PB4<Alternate<AF12>>, PB6<Alternate<AF12>>]
+    D1: [PA8<Alternate<AF12>>]
+    D2: [PA9<Alternate<AF12>>]
+    D3: [PB5<Alternate<AF12>>]
+}
+
+#[derive(Copy, Clone)]
+pub enum Buswidth {
+    Buswidth1 = 0,
+    Buswidth4 = 1,
+}
+
+enum PowerCtrl {
+    Off = 0b00,
+    On = 0b11,
+}
+
+#[allow(dead_code)]
+enum ClockFreq {
+    F24Mhz = 0,
+    F16Mhz = 1,
+    F12Mhz = 2,
+    F8Mhz = 8,
+    F4Mhz = 10,
+    F1Mhz = 46,
+    F400Khz = 118,
+}
+
+#[repr(u32)]
+#[allow(dead_code)]
+enum CmdAppOper {
+    VOLTAGE_WINDOW_SD = 0x8010_0000,
+    HIGH_CAPACITY = 0x4000_0000,
+    SDMMC_STD_CAPACITY = 0x0000_0000,
+    SDMMC_CHECK_PATTERN = 0x0000_01AA,
+    SD_SWITCH_1_8V_CAPACITY = 0x0100_0000,
+}
+
+#[derive(Eq, PartialEq, Copy, Clone)]
+pub enum Response {
+    None = 0,
+    Short = 1,
+    Long = 3,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum CardVersion {
+    V2 = 2,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum CardType {
+    /// Standard Capacity (< 2Gn)
```

Comment:
Gb

Reply:
> Great work!
> 
> The example is a bit complicated to be followed as an example, but I suppose that's your test application.

Thanks!

Not really my test program, but when I finished the driver someone made a usb mass storage crate, and I couldn't resist using it. I agree it's a bit to complicated. I could make a simpler example with just status and block read and write and have the more complex examples in my feather-board crate.